[WIP/RFC] Add missing key to wildmenu

erw7 opened the pull request:
Add missing keymap to wildmenu popup menu compared to insert mode popup menu.

Ref. #10693

- [x]  Assign `<Left>` and `<Right>` to directory moves.
- [x]  Fix the problem that the item selected by `<Ctrl-Y>` is not reflected when the item is moved by `<Up>`, `<Down>`.
- [ ] Test
- [ ] Document

erw7 committed:
Add missing key to wildmenu

Add missing keymap to wildmenu popup menu compared to insert mode popup
menu

justinmk commented:
Looks nice! Let's wait until 0.5 (releasing 0.4 soon).

bfredl commented:
This seems to break the existing `<up>` and `<down>`' wildmenu keys ( up and down in the directory hierarchy, try with `:e some/nested/dir` ). Perhaps there should be an option (or runtime bindings) for swapped mode, where this functionality instead is on `<left>` and `<right>`, when up/down/pgup/pgdown is used for the pum-style wildmenu.

erw7 committed:
Change directory move key on wop=pum

erw7 committed:
Fix problem that Ctrl_Y doesn't work on wop=pum

erw7 committed:
Fix behavior of <PageUP>, <PageDown>, <Up> and <Down> keys

erw7 committed:
Fix test

Change the test according to the keymap change when wildoptions=pum.

erw7 committed:
Fix memory leak

erw7 committed:
Fix document

erw7 committed:
Fix a test that forgot to fix

erw7 committed:
Fix access outside array range

Fix the problem of accessing outside the array range, if <C-y> is used
in the non-selected state of the popupmenu.

erw7 committed:
Add test case for <C-y>

erw7 committed:
Fix warning for test

Add intermediate state to screen test.

howardcochran commented:
Since this is an RFC, I'll comment that I really like this feature. The menu is oriented vertically, so the existing mapping messes with my brain! That this would be different from the keys used for the horizontal wildmenu seems OK since this menu is oriented vertically.  But it would be fine to make it optional.

Incidentally, I tried to correct the mapping myself with a mapping like this (based on text in the help for ins-completion-menu, but it turns out that pumvisible() returns false when the pum-style wildmenu is visible:
cnoremap <Down> <C-R>=pumvisible() ? "\<lt>Right>" : "\<lt>Down>"<CR>

So, I hope erw7's change can be accepted.

bfredl commented:
>  but it turns out that pumvisible() returns false when the pum-style wildmenu is visible

We need to fix it then. BTW, with `<expr>` mapping `<lt>Right>` shouldn't be needed, just `<right>` is fine.

In general, this is an improvement for new users, but changing the builtin bindings by default might break existing user's muscle memory, and, perhaps more fundamentally, also existing mappings (and perhaps plugins feeding the keys). For instance, I have a mapping from `/` to `<down>` when wildmenu is active to expand the inner dir. Part of the motivation to enable `wop=pum` per default is that it is a visual only change, and won't break existing plugins or configs (of course some vim users consider changing visuals is already "breaking" their config, but breaking mappings and behaviors is still even worse).

bfredl commented:
> 	CTRL-Y		- Accept the currently selected match and stop completion.
	CTRL-E		- End completion, go back to original typed text.

Also there are not related to orientation, and shouldn't be conditioned on `wop=pum`. (I e we should either apply then unconditionally, or by separate option).

erw7 committed:
Change to optionally assign keys

erw7 committed:
Change Ctrl and Ctrl Y to work without pum

erw7 commented:
>  but it turns out that pumvisible() returns false when the pum-style wildmenu is visible:

@howardcochran  It seems to work with the following settings at my hand. Therefore, `pumvisible()` returns `true` if the pum-style wildmenu is displayed.

```vim
cnoremap <expr><Up> pumvisible() ? "\<Up>" : "\<Down>"
cnoremap <expr><Down> pumvisible() ? "\<Right>" : "\<Down>"
```

erw7 commented:
@bfredl I was most worried about which function was assigned to which key during implementation. The reasons for the current assignment are:

- The behavior should be intuitive.
- Those that look the same (complement pum and command line pum) should be capable of the same operation.
- The default settings should be easy for beginners to understand (If you are an expert, changing the settings should be easy.).

The code to change the key using `compl_match_array` is complicated, so I assigned the key according to the option, but for the above reasons, the default assignment is left as it is.

erw7 commented:
>> CTRL-Y - Accept the currently selected match and stop completion.
>> CTRL-E - End completion, go back to original typed text.
>
> Also there are not related to orientation, and shouldn't be conditioned on wop=pum. (I e we should either apply then unconditionally, or by separate option).

Changed to be able to use `CTRL-Y` and `CTRL-E` even if `wop=pum` is not set. Although not intuitive, I think it would be useful to be able to use `PageUP` and `PageDown` even if `wop=pum` is not set, just like `Ctrl-Y` and `Ctlr-E`. Is it possible to think of a better implementation method than the implementation that keeps the number displayed in `win_redr_status_match()` in the static variable like `pum_height`?

bfredl commented:
> The default settings should be easy for beginners to understand (If you are an expert, changing the settings should be easy.).

I do not think this strict dichotomy between beginner and expert is very useful. If you know a little about insert mode completion, it might already be initiuitive for you to prefer the CTRL-N and CTRL-P keys to navigate the wildmenu (or just hit 'wildchar' multiple times), which are much closer to the home row. Then you find navigating to a subdir is a clumsy, so you read the `:help` about wildmenu and find that `<down>` key can used to expand a subdir. If you use this frequently, you might end up doing something like

    cnoremap <c-XX> <down>

Even if you don't yet know about `<expr>` mappings, chances are you can find a key XX you won't use anyway. So it is not uncommon to develop the intuition that the meaning of a key is associated with the _name_ of the key, rather than its physical location, and then it is weird that a mapping broke just because the same functional element (wildmenu) changed its visual appearance.

Now we can end up consider breaking these mappings "worth it", for improving the out-of-box experience for some other users, but it is not certainly not just "expert" configs that get broken.

howardcochran commented:
OK, I agree with @bfredl 's arguments for not changing the default mapping of the arrow keys since it could break existing scripts or mappings, especially since wildoption=pum,... was is the default.  However, I do think some verbage in the help about how to change these using <expr> mappings would be helpful.  Would a patch from me to that effect be welcome?

howardcochran commented:
> > but it turns out that pumvisible() returns false when the pum-style wildmenu is visible:
> 
> @howardcochran It seems to work with the following settings at my hand. Therefore, `pumvisible()` returns `true` if the pum-style wildmenu is displayed.
> 
> ```viml
> cnoremap <expr><Down> pumvisible() ? "\<Right>" : "\<Down>"
> ```

So it is when using <expr> mappings, thank you!

howardcochran commented:
> ... so you read the `:help` about wildmenu and find that `<down>` key can used to expand a subdir. If you use this frequently, you might end up doing something like
> 
> ```
> cnoremap <c-XX> <down>
> ```
> 
> Even if you don't yet know about `<expr>` mappings, chances are you can find a key XX you won't use anyway.

Actually, I think there might be an issue here, specifically with a mapping to <down> while the new pum wildmenu is visible!  I set up these mapping to make the arrow keys follow the geometry of the menu for me:
```viml
cnoremap <expr> <Up>    pumvisible() ? "\<Left>"  : "\<Up>"
cnoremap <expr> <Down>  pumvisible() ? "\<Right>" : "\<Down>"
cnoremap <expr> <Left>  pumvisible() ? "\<Up>"    : "\<Left>"
cnoremap <expr> <Right> pumvisible() ? "\<Down>"  : "\<Right>"
```
The mapping for the first three work correctly, but the one for <Right> does not. When the menu is visible, this mapping from <Right> to <Down> dismisses the menu and inserts a literal Ctrl-I (^I) into the command buffer!

As a quick test, I made this simpler mapping:
```viml
cnoremap / <Down>
```
Sure enough, when the pum-style wildmenu is visible, pressing the slash key cancels the menu and inserts ^I.  But pressing the slash key when the menu is not visible moves downward in the command history as expected.

I am using release 4.2.  Can you reproduce this? Should I open this as a separate issue?  Thanks.

bfredl commented:
Seems `<down>` is implemented by restarting wildmenu after inserting the separator:
```
   s->c = (int)p_wc;
```
This however only works reliably if wildcharm is equal to wildchar, as wildchar doesn't work in macros, noremap etc. I think we can fix this by changing it to 
```
   s->c = (int)p_wcm;
```
As 'wildcharm' happen to work both in "mapped" and "unmapped" contexts. (Not that I understand why 'wildcharm' ever was added as an option. This functionality should just have been hardcoded to   `<c-z>`, so that `cnoremap whatever <c-z>` always works. There is no other functionality on `:<c-z>` to conflict with).

justinmk commented on runtime/doc/options.txt:
```diff
@@ -6652,6 +6652,24 @@ A jump table for the options with a short description can be found at |Q_op|.
+			  but don't insert it.
+	<Down>		- Select the next match, as if CTRL-N was used, but
+			  don't insert it.
+	<Reight>	- in filename completion: move into a subdirectory.
```
```suggestion
	<Right>	- in filename completion: move into a subdirectory.
```

justinmk commented:
When I glanced at this before, I didn't realize that it changes (breaks) existing keys.

- Main question is: why can't' this be done with mappings? I have never noticed a problem because I use `<c-n>` / `<c-p>` (which is much better than arrow keys...), but if users are really bothered by this we can just provide a snippet in the help docs that they can paste into their config.  
- Is 'wildpumkey' still needed then? Seems like it just changes the behavior of arrow keys.
- This is a lot of code just to avoid some mappings. What am I missing, that isn't fixed by mappings?

bfredl commented:
@justinmk I think nothing more than the one-line change above, then it should be easily remapped. we could also consider freezing 'wildcharm' at the same time, though might break some configs that change it for whatever reason. I could do a github code search at least to see if exposed vimrc's ever sets 'wildcharm'/'wcm' to anything other than `<c-z>`.
